flink-connectors/flink-sql-connector-hive-1.2.2/src/main/resources/META-INF/NOTICE

```diff
@@ -8,6 +8,41 @@ This project bundles the following dependencies under the Apache Software Licens
 
 - org.apache.hive:hive-exec:1.2.2
 - org.apache.hive:hive-metastore:1.2.2
+- org.apache.hive:hive-common:1.2.2
+- org.apache.hive:hive-serde:1.2.2
+- org.apache.hive.shims:hive-shims-0.20S:1.2.2
```

Comment:
How did you determine that these are the included dependencies?

Reply:
I checked the pom file of hive-exec to determine what artifacts are included. Is that the right way to do it?

Reply:
yes, that's good. It would be good to document this _somewhere_, maybe the [Licensing](https://cwiki.apache.org/confluence/display/FLINK/Licensing) wiki page until we found a better place.

Reply:
Isn't it better to use mvn dependency:tree to check the included artifacts to also see the transitive dependencies of hive-exec?

Reply:
For reference 
```
[INFO] --- maven-dependency-plugin:3.1.1:tree (default-cli) @ flink-sql-connector-hive-1.2.2_2.11 ---
[INFO] org.apache.flink:flink-sql-connector-hive-1.2.2_2.11:jar:1.12-SNAPSHOT
[INFO] +- org.apache.flink:flink-connector-hive_2.11:jar:1.12-SNAPSHOT:compile
[INFO] +- org.apache.hive:hive-exec:jar:1.2.2:compile
[INFO] |  +- org.apache.hive:hive-ant:jar:1.2.2:compile
[INFO] |  |  \- org.apache.velocity:velocity:jar:1.5:compile
[INFO] |  |     \- oro:oro:jar:2.0.8:compile
[INFO] |  +- org.apache.hive:hive-shims:jar:1.2.2:compile
[INFO] |  |  +- org.apache.hive.shims:hive-shims-common:jar:1.2.2:compile
[INFO] |  |  +- org.apache.hive.shims:hive-shims-0.20S:jar:1.2.2:runtime
[INFO] |  |  +- org.apache.hive.shims:hive-shims-0.23:jar:1.2.2:runtime
[INFO] |  |  |  \- org.apache.hadoop:hadoop-yarn-server-resourcemanager:jar:2.6.0:runtime
[INFO] |  |  |     +- com.google.inject.extensions:guice-servlet:jar:3.0:runtime
[INFO] |  |  |     +- com.google.inject:guice:jar:3.0:runtime
[INFO] |  |  |     |  +- javax.inject:javax.inject:jar:1:runtime
[INFO] |  |  |     |  \- aopalliance:aopalliance:jar:1.0:runtime
[INFO] |  |  |     +- com.sun.jersey:jersey-json:jar:1.9:runtime
[INFO] |  |  |     |  +- com.sun.xml.bind:jaxb-impl:jar:2.2.3-1:runtime
[INFO] |  |  |     |  +- org.codehaus.jackson:jackson-jaxrs:jar:1.8.3:runtime
[INFO] |  |  |     |  \- org.codehaus.jackson:jackson-xc:jar:1.8.3:runtime
[INFO] |  |  |     +- com.sun.jersey.contribs:jersey-guice:jar:1.9:runtime
[INFO] |  |  |     +- org.apache.hadoop:hadoop-yarn-common:jar:2.4.1:runtime
[INFO] |  |  |     +- org.apache.hadoop:hadoop-yarn-api:jar:2.6.0:runtime
[INFO] |  |  |     +- javax.xml.bind:jaxb-api:jar:2.3.1:runtime
[INFO] |  |  |     |  \- javax.activation:javax.activation-api:jar:1.2.0:runtime
[INFO] |  |  |     +- org.codehaus.jettison:jettison:jar:1.1:runtime
[INFO] |  |  |     +- com.sun.jersey:jersey-client:jar:1.9:runtime
[INFO] |  |  |     +- org.mortbay.jetty:jetty-util:jar:6.1.26:runtime
[INFO] |  |  |     +- org.apache.hadoop:hadoop-yarn-server-common:jar:2.6.0:runtime
[INFO] |  |  |     |  \- org.fusesource.leveldbjni:leveldbjni-all:jar:1.8:runtime
[INFO] |  |  |     +- org.apache.hadoop:hadoop-yarn-server-applicationhistoryservice:jar:2.6.0:runtime
[INFO] |  |  |     \- org.apache.hadoop:hadoop-yarn-server-web-proxy:jar:2.6.0:runtime
[INFO] |  |  |        \- org.mortbay.jetty:jetty:jar:6.1.26:runtime
[INFO] |  |  \- org.apache.hive.shims:hive-shims-scheduler:jar:1.2.2:runtime
[INFO] |  +- commons-codec:commons-codec:jar:1.10:compile
[INFO] |  +- commons-httpclient:commons-httpclient:jar:3.0.1:compile
[INFO] |  +- commons-io:commons-io:jar:2.4:compile
[INFO] |  +- commons-logging:commons-logging:jar:1.1.3:compile
[INFO] |  +- log4j:apache-log4j-extras:jar:1.2.17:compile
[INFO] |  +- org.antlr:antlr-runtime:jar:3.4:compile
[INFO] |  |  +- org.antlr:stringtemplate:jar:3.2.1:compile
[INFO] |  |  \- antlr:antlr:jar:2.7.7:compile
[INFO] |  +- org.antlr:ST4:jar:4.0.4:compile
[INFO] |  +- org.apache.ant:ant:jar:1.9.1:compile
[INFO] |  |  \- org.apache.ant:ant-launcher:jar:1.9.1:compile
[INFO] |  +- org.apache.commons:commons-compress:jar:1.20:compile
[INFO] |  +- org.apache.ivy:ivy:jar:2.4.0:compile
[INFO] |  +- org.apache.zookeeper:zookeeper:jar:3.4.14:compile
[INFO] |  |  +- com.github.spotbugs:spotbugs-annotations:jar:3.1.9:compile
[INFO] |  |  \- org.apache.yetus:audience-annotations:jar:0.5.0:compile
[INFO] |  +- org.apache.curator:curator-framework:jar:2.6.0:compile
[INFO] |  |  \- org.apache.curator:curator-client:jar:2.6.0:compile
[INFO] |  +- org.apache.curator:apache-curator:pom:2.6.0:compile
[INFO] |  +- org.codehaus.groovy:groovy-all:jar:2.1.6:compile
[INFO] |  +- org.datanucleus:datanucleus-core:jar:3.2.10:compile
[INFO] |  +- org.apache.calcite:calcite-core:jar:1.2.0-incubating:compile
[INFO] |  |  +- org.apache.calcite:calcite-linq4j:jar:1.2.0-incubating:compile
[INFO] |  |  +- net.hydromatic:eigenbase-properties:jar:1.1.5:compile
[INFO] |  |  +- org.codehaus.janino:janino:jar:2.7.6:compile
[INFO] |  |  \- org.codehaus.janino:commons-compiler:jar:2.7.6:compile
[INFO] |  +- org.apache.calcite:calcite-avatica:jar:1.2.0-incubating:compile
[INFO] |  +- stax:stax-api:jar:1.0.1:compile
[INFO] |  \- jline:jline:jar:2.12:compile
[INFO] +- org.apache.hive:hive-metastore:jar:1.2.2:compile
[INFO] |  +- org.apache.hive:hive-serde:jar:1.2.2:compile
[INFO] |  |  +- org.apache.hive:hive-common:jar:1.2.2:compile
[INFO] |  |  |  +- joda-time:joda-time:jar:2.5:compile
[INFO] |  |  |  \- org.json:json:jar:20090211:compile
[INFO] |  |  +- org.apache.avro:avro:jar:1.8.2:compile
[INFO] |  |  |  +- org.codehaus.jackson:jackson-mapper-asl:jar:1.9.13:compile
[INFO] |  |  |  +- com.thoughtworks.paranamer:paranamer:jar:2.7:compile
[INFO] |  |  |  +- org.xerial.snappy:snappy-java:jar:1.1.4:compile
[INFO] |  |  |  \- org.tukaani:xz:jar:1.5:compile
[INFO] |  |  +- net.sf.opencsv:opencsv:jar:2.3:compile
[INFO] |  |  \- com.twitter:parquet-hadoop-bundle:jar:1.6.0:compile
[INFO] |  +- com.google.guava:guava:jar:14.0.1:compile
[INFO] |  +- com.jolbox:bonecp:jar:0.8.0.RELEASE:compile
[INFO] |  +- commons-cli:commons-cli:jar:1.3.1:compile
[INFO] |  +- commons-lang:commons-lang:jar:2.6:compile
[INFO] |  +- org.apache.derby:derby:jar:10.10.2.0:compile
[INFO] |  +- org.datanucleus:datanucleus-api-jdo:jar:3.2.6:compile
[INFO] |  +- org.datanucleus:datanucleus-rdbms:jar:3.2.9:compile
[INFO] |  +- commons-pool:commons-pool:jar:1.5.4:compile
[INFO] |  +- commons-dbcp:commons-dbcp:jar:1.4:compile
[INFO] |  +- javax.jdo:jdo-api:jar:3.0.1:compile
[INFO] |  |  \- javax.transaction:jta:jar:1.1:compile
[INFO] |  \- org.apache.thrift:libthrift:jar:0.9.2:compile
[INFO] |     +- org.apache.httpcomponents:httpclient:jar:4.5.3:compile
[INFO] |     \- org.apache.httpcomponents:httpcore:jar:4.4.6:compile
[INFO] +- org.apache.thrift:libfb303:jar:0.9.2:compile
[INFO] +- org.apache.orc:orc-core:jar:nohive:1.4.3:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:2.5.0:compile
[INFO] |  +- org.apache.hadoop:hadoop-common:jar:2.4.1:compile
[INFO] |  |  +- org.apache.hadoop:hadoop-annotations:jar:2.4.1:compile
[INFO] |  |  +- org.apache.commons:commons-math3:jar:3.5:compile
[INFO] |  |  +- xmlenc:xmlenc:jar:0.52:compile
[INFO] |  |  +- commons-net:commons-net:jar:3.1:compile
[INFO] |  |  +- commons-collections:commons-collections:jar:3.2.2:compile
[INFO] |  |  +- com.sun.jersey:jersey-core:jar:1.9:compile
[INFO] |  |  +- com.sun.jersey:jersey-server:jar:1.9:compile
[INFO] |  |  |  \- asm:asm:jar:3.1:compile
[INFO] |  |  +- commons-configuration:commons-configuration:jar:1.7:compile
[INFO] |  |  |  \- commons-beanutils:commons-beanutils:jar:1.8.3:compile
[INFO] |  |  +- org.codehaus.jackson:jackson-core-asl:jar:1.8.8:compile
[INFO] |  |  +- org.apache.hadoop:hadoop-auth:jar:2.4.1:compile
[INFO] |  |  \- com.jcraft:jsch:jar:0.1.42:compile
[INFO] |  \- org.apache.hive:hive-storage-api:jar:2.2.1:compile
[INFO] +- io.airlift:aircompressor:jar:0.8:compile
[INFO] |  \- io.airlift:slice:jar:0.29:compile
[INFO] |     \- org.openjdk.jol:jol-core:jar:0.2:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.15:provided
[INFO] +- com.google.code.findbugs:jsr305:jar:1.3.9:provided
[INFO] +- org.apache.flink:flink-test-utils-junit:jar:1.12-SNAPSHOT:test
[INFO] +- org.apache.flink:force-shading:jar:1.12-SNAPSHOT:compile
[INFO] +- junit:junit:jar:4.12:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] +- org.mockito:mockito-core:jar:2.21.0:test
[INFO] |  +- net.bytebuddy:byte-buddy:jar:1.8.15:test
[INFO] |  +- net.bytebuddy:byte-buddy-agent:jar:1.8.15:test
[INFO] |  \- org.objenesis:objenesis:jar:2.1:test
[INFO] +- org.powermock:powermock-module-junit4:jar:2.0.4:test
[INFO] |  \- org.powermock:powermock-module-junit4-common:jar:2.0.4:test
[INFO] |     +- org.powermock:powermock-reflect:jar:2.0.4:test
[INFO] |     \- org.powermock:powermock-core:jar:2.0.4:test
[INFO] |        \- org.javassist:javassist:jar:3.24.0-GA:test
[INFO] +- org.powermock:powermock-api-mockito2:jar:2.0.4:test
[INFO] |  \- org.powermock:powermock-api-support:jar:2.0.4:test
[INFO] +- org.hamcrest:hamcrest-all:jar:1.3:test
[INFO] +- org.apache.logging.log4j:log4j-slf4j-impl:jar:2.12.1:test
[INFO] +- org.apache.logging.log4j:log4j-api:jar:2.12.1:test
[INFO] +- org.apache.logging.log4j:log4j-core:jar:2.12.1:test
[INFO] \- org.apache.logging.log4j:log4j-1.2-api:jar:2.12.1:test
[INFO] ------------------------------------------------------------------------
```

Reply:
I believe not all dependencies reported by `mvn dependency:tree` are included?

Reply:
And some dependencies included may not be visible in the dependency tree because of the dependency-reduced pom

Reply:
No, not all of them, but I believe some (like the hadoop dependencies from the shims)

Reply:
I also don't know what's the best way forward. The problem is that the Hive project is not providing a proper `NOTICE` file that we could use.

We could do something crazy and build hive 1.2.2 ourselves, and take the output of their maven-shade-plugin run in the "hive/ql" (=`hive-exec`) run to get all the included dependencies.

Reply:
The hive-exec pom includes a very explicit shade-plugin include configuration;
```
                  <!-- order is meant to be the same as the ant build -->
                  <include>org.apache.hive:hive-common</include>
                  <include>org.apache.hive:hive-exec</include>
                  <include>org.apache.hive:hive-serde</include>
                  <include>com.esotericsoftware.kryo:kryo</include>
                  <include>com.twitter:parquet-hadoop-bundle</include>
                  <include>org.apache.thrift:libthrift</include>
                  <include>commons-lang:commons-lang</include>
                  <include>org.apache.commons:commons-lang3</include>
                  <include>org.jodd:jodd-core</include>
                  <include>org.json:json</include>
                  <include>org.apache.avro:avro</include>
                  <include>org.apache.avro:avro-mapred</include>
                  <include>org.apache.hive.shims:hive-shims-0.20S</include>
                  <include>org.apache.hive.shims:hive-shims-0.23</include>
                  <include>org.apache.hive.shims:hive-shims-0.23</include>
                  <include>org.apache.hive.shims:hive-shims-common</include>
                  <include>com.googlecode.javaewah:JavaEWAH</include>
                  <include>javolution:javolution</include>
                  <include>com.google.protobuf:protobuf-java</include>
                  <include>org.iq80.snappy:snappy</include>
                  <include>org.codehaus.jackson:jackson-core-asl</include>
                  <include>org.codehaus.jackson:jackson-mapper-asl</include>
                  <include>com.google.guava:guava</include>
                  <include>net.sf.opencsv:opencsv</include>
                  <include>org.apache.hive:spark-client</include>
                  <include>joda-time:joda-time</include>
```

Reply:
Yes, but you don't know what they include transitively (for example hadoop)

Reply:
I don't think transitive dependencies are included, if an explicit `include` configuration is present. But of course, checking include configuration doesn't cover the case where an included dependency is also a uber jar, e.g. Kryo. In that case, seems we need to do this recursively...

Reply:
For that you will naturally have to move up the dependency tree and investigate each dependencies. There's no other way. For each you will either have to check the packaging output, or analyze the build files.

Traditionally we did not go this far though, because it is ultimately impractical to do this for all dependencies.